contribution/translation/prevalidator.php

```diff
@@ -67,282 +64,98 @@ public function get_helper()
-			else if (strpos($dir, $types['prosilver']) != 0)
-			{
-				$dir_prefix = explode($types['prosilver'] ,$dir);
-			}
-			else if (strpos($dir, $types['subsilver2']) != 0)
-			{
-				$dir_prefix = explode($types['subsilver2'] ,$dir);
-			}
-			$dir_clean = str_replace('/'.$iso_code.'/', '/en/', str_replace($dir_prefix[0], '', $dir));
 
-			foreach ($files as $file)
-			{
-				$list_uploaded_files[] = $dir_clean . $file;
-			}
-		}
-		// It's time to check if each file uploaded in the package is allowed
-		foreach ($list_uploaded_files as $file)
+		else
 		{
-			if (!in_array($file, $list_reference_files) && !in_array($file, $this->ignore_files))
-			{
-				$error[] = $this->user->lang('WRONG_FILE', str_replace('/en/', '/'.$iso_code.'/', $file)); // report a wrong file
-			}
-		}
+			// We don't need to clean up (delete) the temporary files here because that is
+			// handled at revision.php:752
+			$package->ensure_extracted();
+			$path = $package->get_temp_path();
 
-		if (!sizeof($error))
-		{
-			$package->repack(); // we have made changes to the package, so replace the original zip file
-		}
-		return $error;
-	}
+			// Rename the extracted directory to be the ISO code.
+			$finder = new Finder();
+			$iterator = $finder->directories()->in($path)->depth('== 0');
 
-	/**
-	 * Compares two phpBB 3.0.x language files and computes the missing keys in the uploaded file
-	 * Does not include the uploaded file for security and uses string check on the file contents
-	 *
-	 * @param string $uploaded_file
-	 * @param string $reference_file
-	 * @return array returns an array with the missing keys
-	 */
-	private function check_missing_keys($reference_file, $uploaded_file)
-	{
-		// Check original file by including the language entries...
-		$lang = $missing_keys = array();
-
-		$contents = file_get_contents($uploaded_file);
-
-		include($reference_file);
-		//$lang_keys = $this->multiarray_keys($lang);
-
-		// General cleanup
-		$file_data = trim(str_replace(array("\r", "\t", ' '), '', $contents));
-		$file_data = explode("\n", $file_data);
-
-		$file_data = str_replace("\n", '', implode("\n", $file_data));
-
-		// Now we have all array keys... let us have a look within $contents if all keys are there...
-		// This can take a bit because we check every key... luckily with strpos...
-		foreach (array_keys($lang) as $current_key)
-		{
-			if (is_string($current_key) && !is_array($lang[$current_key]) && strpos($file_data, "'{$current_key}'=>") === false)
+			if ($iterator->count() == 1)
 			{
-				if (empty($lang[$current_key]))
-				{
-					continue;
-				}
+				$iterator = $finder->getIterator();
+				$iterator->rewind();
 
-				$missing_keys[] = $current_key;
+				// We know there's only one result
+				$root_extracted = $iterator->current();
 			}
-		}
 
-		return $missing_keys;
-	}
+			// Rename the directory to use the iso code
+			$new_directory_name = str_replace($root_extracted->getFilename(), $origin_iso, $root_extracted->getPathname());
```

Comment:
`$root_extracted` won't exist if the above condition was false.

Reply:
Done @ 8fb5aec912a6dacb7db05a20dd590ae5253f2cc1 (rearranged the exception handling to cover this)